### DISCUSSION TITLE
how to impose the deformation output by abaqus to Moose

### DISCUSSION THREAD

**Discussion by HairyLittleBoy** [idaholab/moose#27836] [Q&A Modules: Solid mechanics] [answered]
Hi

I build a cubic volume in Moose and want it to deform the way same as the simulation result obtained with ABAQUS, which means the displacements of each surface (or vertex) of the cubic be the same as ABAQUS. 

I directly give the displacement to each surface like:
```
[BCs]
  [tdisp1]
    type = DirichletBC
    variable = disp_x
    boundary = bottom
    value = 0.003
  []
  [tdisp2]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0236
  []
  [tdisp3]
    type = DirichletBC
    variable = disp_z
    boundary = bottom
    value = 0.0236
  []


  [tdisp4]
    type = DirichletBC
    variable = disp_x
    boundary = top
    value = 0.0236
  []
  [tdisp5]
    type = DirichletBC
    variable = disp_y
    boundary = top
    value = 0.0236
  []
  [tdisp6]
    type = DirichletBC
    variable = disp_z
    boundary = top
    value = 0.0236
  []  
 
  [tdisp7]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0.0302
  []
  [tdisp8]
    type = DirichletBC
    variable = disp_y
    boundary = left
    value =  0.003
  []
  [tdisp9]
    type = DirichletBC
    variable = disp_z
    boundary = left
    value =  -0.0112
  [] 

  [tdisp10]
    type = DirichletBC
    variable = disp_x
    boundary = right
    value = 0.003
  []
  [tdisp11]
    type = DirichletBC
    variable = disp_y
    boundary = right
    value = 0.003
  []
  [tdisp12]
    type = DirichletBC
    variable = disp_z
    boundary = right
    value = 0.003
  [] 


  [tdisp13]
    type = DirichletBC
    variable = disp_x
    boundary = back
    value = 0.0322
  []
  [tdisp14]
    type = DirichletBC
    variable = disp_y
    boundary = back
    value = 0.0322
  []
  [tdisp15]
    type = DirichletBC
    variable = disp_z
    boundary = back
    value = -0.0132
  [] 

  
  [tdisp16]
    type = DirichletBC
    variable = disp_x
    boundary = front
    value = 0.0259
  []
  [tdisp17]
    type = DirichletBC
    variable = disp_y
    boundary = front
    value = 0.0259
  []
  [tdisp18]
    type = DirichletBC
    variable = disp_z
    boundary = front
    value =  0.0259
  [] 
[]
```
however, it always does not converge at iteration 0. The same situation when I prescibe the displacement to the 8 vertexes

Also, if the deformation gradient can be defined, it might also works, I found following discussion but still not understande how to do, may need some details.
[https://github.com/idaholab/moose/discussions/19487](url)

**Comment by GiudGiud** [comment]
Hello

Can you set preset=false to all these boundary conditions and see if it converges?

Guillaume

**Reply by HairyLittleBoy** [reply]
thank you for your reply. No it doesn't work. I am applying a crystal plastic simulation of pure Ti in ABAQUS by using a UMAT without twinning mechanism, so I'm trying to export the deformation from ABAQUS result and give them to Moose. However, my boundary condition setting leads to divergence even I set preset=false. I think I will do my CP simulation of indentation totally within Moose. The discussions #16716, the input example (modules/contact/examples/3d_berkovich/indenter_berkovich_friction.i) and the vedios by Nicolò Grilli (https://www.youtube.com/watch?v=qG7EoX34hZ8) may help.

**Reply by HairyLittleBoy** [reply]
Also, after I read the modules/contact/examples/3d_berkovich/indenter_berkovich_friction.i, I cannot find the definition of crystal orientation nowhere, how can I give each grain an orientation (through bunger euler angle or orientation matrix) ? @GiudGiud

**Reply by HairyLittleBoy** [reply]
after a few searching, I found this :
```
/solid_mechanics/src/userobjects/EulerAngleFileReader.C
```
but what I'm interested in is the modified Kalidindi model for hcp : 
```
[Materials]
    [elasticity_tensor]
       type = ComputeElasticityTensorCP
       C_ijkl = '1.622e5 9.18e4 6.88e4 1.622e5 6.88e4 1.805e5 4.67e4 4.67e4 4.67e4' 
    fill_method = symmetric9
  []
  [stress]
    type = ComputeMultipleCrystalPlasticityStress
    crystal_plasticity_models = 'twin_xtalpl'
    tan_mod_type = exact
  []
  [twin_xtalpl]
    type = CrystalPlasticityTwinningKalidindiUpdate
    base_name = twin
    crystal_lattice_type = HCP
    unit_cell_dimension = '2.934e-7 2.934e-7 4.657e-7' 
    number_slip_systems = 6
    slip_sys_file_name = 'hcp_tensile_twin_systems.txt'
    initial_twin_lattice_friction = 1140
    non_coplanar_coefficient_twin_hardening = 10000
    coplanar_coefficient_twin_hardening = 1000
    characteristic_twin_shear = 0.167
  []
[]
```
CrystalPlasticityTwinningKalidindiUpdate is under [Material] instead of [UserObjects], Can I apply the "EulerAngleFileReader" to define the orientation of crystal ?

**Reply by GiudGiud** [reply]
Hello

The reader is a user object because that's all it does, read the angles from a file iirc. 
Now the materials have to pull in the UserObject (using getUserObject on the problem) to query the angles. 

If this has not been implemented for Kalindi, then you'll need to implement that

**Reply by sapitts** [reply, marked as answer]
Use the parameters in `ComputeElasticityTensorCP` to provide the euler angles, options include:
 - `euler_angle_1`, `euler_angle_2`, and `euler_angle_3`
 - `euler_angle_variables`
 - `read_prop_user_object`, or
 - `rotation_matrix`
See the many example input files listed at the bottom of the documentation page for how to use one of these options: https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeElasticityTensorCP.html

If you are looking for a HCP specific constitutive model, I recommend the [CrystalPlasticityHCPDislocationSlipBeyerleinUpdate](https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.html). The twinning Kalidindi model with the HCP lattice is intended only for regression testing purposes

**Reply by HairyLittleBoy** [reply]
thank you @sapitts , I will try that !
